rfcs/2020-10-15-3480-file-source-rework.md

```diff
@@ -0,0 +1,425 @@
+# RFC 3480 - 2020-10-15 - File Source Rework
+
+The `file` source is one of Vector's oldest and most widely used components.
+While it began with relatively simple interface and implementation, it has
+collected a large number of knobs, warts, and general issues over time. To
+ensure that this important source continues to serve users well, we should
+invest effort into making it easier to use, extend, and debug.
+
+## Scope
+
+This RFC will cover a general overhaul of the `file` source internals and
+user-facing config. The focus will be on migrating the existing implementation
+to a better structure, not rewriting from scratch.
+
+## Motivation
+
+There are three pages of issues tagged `source: file` in the Vector repo. Below
+I've tried to collect some of the more notable ones and categorize them:
+
+* File source retains files on disk ([#763](https://github.com/timberio/vector/issues/763))
+
+  * Slowness due to many files
+  * Need sequential instead of fair reads
+  * Unclear why files handles were open
+  * Unclear that files were not done being processed
```

Comment:
I think this issue might also suffer the same issue in https://github.com/timberio/vector/pull/4609/files#r506955416.

* Unclear why file handles for rotated files were open.